### ISSUE TITLE
handling of prefab structs is totally wrong

### ISSUE THREAD

**Issue by pnwamk** [racket/typed-racket#621]
```racket
#lang typed/racket/base
 
(struct fubar ([id : Byte]) #:prefab)
 
(define rval : Any (read (open-input-string "#s(fubar NotANumber)")))
 
 
(when (fubar? rval)
  (displayln (+ 1 (fubar-id rval))))
```

The short of the issue, prefab structs are more like `cons` than they are standard non-prefab structs since it is impossible to guarantee only certain types will inhabit their fields. 

My thoughts currently are we should treat prefabs like named polymorphic tuples. More discussion/thought on this probably required before a real solution can be attempted.

**Comment by samth** [comment]
There are some other issues open related to prefabs, I think. But this is the right plan.

**Comment by pnwamk** [comment]
Fixed by 7599aedf2c8ea354d1b4965ec129ea4dd85f8010.
